### ISSUE TITLE
Potential Issue When Tiling to a GlobalLayout

### ISSUE THREAD

**Issue by jbouffard** [geotrellis/geotrellis-contrib#41]
# Overview

I've been running into an issue in GPS, and the problem might be from VLM. The issue is that Tiles produced using code from `geotrellis.contrib.vlm` are not the expected result.

## Background

In GPS, I'm trying to implement a [read_to_layout](https://github.com/jbouffard/geopyspark/blob/feature/read_to_layout/geopyspark/geotrellis/layer.py#L1048) method that takes a list of sources and tiles them to a given layout as well as potentially reproject the data. Right now, the different layouts that can be given are: `LocalLayout` or `GlobalLayout`. `LocalLayout` has been working without issue, however, `GlobalLayout` has been producing incorrect Tiles.

## The Code

The Scala code for `read_to_layout` can be found [here](https://github.com/jbouffard/geopyspark/blob/feature/read_to_layout/geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/vlm/RasterSource.scala#L102). In this code, a reprojection is performed on the given `RasterSource`s if specified, the data is then tiled to the given layout, and the resulting data is read in/formatted from the layout's specifications.

## The Tests

The tests for this feature were written in Python and can be found [here](https://github.com/jbouffard/geopyspark/blob/feature/read_to_layout/geopyspark/tests/geotrellis/tiled_layer_tests/tiled_raster_layer_test.py#L16). In this test, the `expected` values are created from GeoTrellis logic.

### The Results

Due to the nature of the Python testing framework, the resulting failure messages aren't too helpful, but here is an example of one:

```
____________________________________ ____________________________________ TiledRasterLayerTest.test_read_no_reproject_global_geotrellis ____________________________________

self = <geopyspark.tests.geotrellis.tiled_layer_tests.tiled_raster_layer_test.TiledRasterLayerTest testMethod=test_read_no_reproject_global_geotrellis>

    def test_read_no_reproject_global_geotrellis(self):
>       self.read(GlobalLayout(tile_size=16, zoom=4), ReadMethod.GEOTRELLIS)

geopyspark/tests/geotrellis/tiled_layer_tests/tiled_raster_layer_test.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geopyspark/tests/geotrellis/tiled_layer_tests/tiled_raster_layer_test.py:33: in read
    self.assertTrue((expected[1].cells == actual[1].cells).all())
E   AssertionError: False is not true
```

### The Outputs

In order to provide more useful information, a zip file that contains three different tiffs has been attached to this issue. These three files are: `all-ones.tif`, `actual-output.tif`, and `expected-output.tif`. The first file is the source file and the other two shows the actual and expected output, respectively.

[outputs.zip](https://github.com/geotrellis/geotrellis-contrib/files/2520630/outputs.zip)

**Comment by echeipesh** [comment]
The `expected-output` is all `NODATA` cells which is a little strange.

**Comment by jbouffard** [comment]
@echeipesh That is a bit strange... I'll need to look into that more on Monday.

**Comment by echeipesh** [comment]
The pytest error would be a bit more useful if the `assert` [here](https://github.com/jbouffard/geopyspark/blob/feature/read_to_layout/geopyspark/tests/geotrellis/tiled_layer_tests/tiled_raster_layer_test.py#L33) mentioned what key the test failed.

**Comment by pomadchin** [comment]
how was `expected` generated? 
GeoTiffRasterSource output is a little bit different at the moment: https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/test/scala/geotrellis/contrib/vlm/RasterSourceRDDSpec.scala#L139

GDALRasterSource would generate a different output comparable to GeoTrellis output in general: https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/test/scala/geotrellis/contrib/vlm/RasterSourceRDDSpec.scala#L147

**Comment by jbouffard** [comment]
@pomadchin `expected` was generated by first reading in the file via `HadoopGeoTiffRDD`, and then tiling to a `ZoomedLayoutScheme` at `zoom level` 4 with a a `cellSize` of 128.

**Comment by pomadchin** [comment]
@jbouffard that's why @echeipesh  mentioned the diff, probably it's an expected thing. And it even should not work for GDAL case.

**Comment by jbouffard** [comment]
@pomadchin It fails with GeoTrellis as well and not just GDAL.

**Comment by pomadchin** [comment]
@jbouffard that's what i said. if you'll follow https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/test/scala/geotrellis/contrib/vlm/RasterSourceRDDSpec.scala#L139-L145 and read the comment, i think my posts above would make some sense. can you post there a detailed diff, as it should not work for the GeoTrellis case as well. Also it looks like nodata values as expected is not correct (?) O:

**Comment by jbouffard** [comment]
@pomadchin Oh, I see what you're saying now. That makes sense. I'll try and get a more detailed difference between the actual and expected tiles. Yeah, that was another think I noticed. I think that might be an issue on the GPS side of things.

**Comment by jbouffard** [comment]
@echeipesh @pomadchin Here is the `expected` and `actual` outputs of `read_to_layout` on a different file using GeoTrellis:

[outputs.zip](https://github.com/geotrellis/geotrellis-contrib/files/2526023/outputs.zip)

What's interesting is that the `actual` output has a slightly smaller range of values than the expected, which I don't think I've seen before.

**Comment by pomadchin** [comment]
what means `slightly smaller range`?

**Comment by jbouffard** [comment]
@pomadchin I meant that the range of values between the two were pretty different. But when I look at the ranges again, I'm realizing that I read them wrong the first time. The `expected` has a value range from `1-2090` whereas the values in `actual` range from `4-307`

**Comment by pomadchin** [comment]
@jbouffard is this still an issue?

**Comment by jbouffard** [comment]
@pomadchin Yeah, this is still an issue in GPS. #46 resolves it though

**Comment by pomadchin** [comment]
And this one doesn't resolve it? https://github.com/geotrellis/geotrellis-contrib/pull/49 

Looks like we need to bump up priority there.

**Comment by pomadchin** [comment]
Can you double check that with #46 everything works like you want and with #49 nope?

**Comment by jbouffard** [comment]
@pomadchin Yeah, I can double check and see if #46 is all that's needed.

**Comment by jbouffard** [comment]
@pomadchin Actually, #49 seems to have resolved the issue :+1:

**Comment by jbouffard** [comment]
This issue has been resolved by #49
